homeassistant/components/switcher_kis/config_flow.py

```diff
@@ -2,9 +2,119 @@
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the start of the config flow."""
+        self.discovered_devices = await async_discover_devices()
+
+        return self.async_show_form(step_id="confirm")
+
+    async def async_step_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle user-confirmation of the config flow."""
+        if len(self.discovered_devices) == 0:
+            return self.async_abort(reason="no_devices_found")
+
+        for device_id, device in self.discovered_devices.items():
+            if device.token_needed:
+                _LOGGER.info("Device with ID %s requires a token", device_id)
+                return await self.async_step_credentials()
+        return await self._create_entry()
+
+    async def async_step_credentials(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the credentials step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.username = user_input.get(CONF_USERNAME)
+            self.token = user_input.get(CONF_TOKEN)
+
+            token_is_valid = await validate_token(
+                user_input[CONF_USERNAME], user_input[CONF_TOKEN]
+            )
+            if token_is_valid:
+                return await self._create_entry()
+            errors["base"] = "invalid_auth"
+        else:
+            user_input = {}
+
+        schema = {
+            vol.Required(CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")): str,
+            vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN, "")): str,
+        }
+
+        return self.async_show_form(
+            step_id="credentials", data_schema=vol.Schema(schema), errors=errors
+        )
+
+    async def async_step_reauth(
+        self, user_input: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Handle configuration by re-auth."""
+        self.entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Dialog that informs the user that reauth is required."""
+        errors: dict[str, str] = {}
+        assert self.entry is not None
+
+        if user_input is not None:
+            token_is_valid = await validate_token(
+                user_input[CONF_USERNAME], user_input[CONF_TOKEN]
+            )
+            if token_is_valid:
+                return self.async_update_reload_and_abort(
+                    self.entry, data={**self.entry.data, **user_input}
+                )
+            errors["base"] = "invalid_auth"
+
+        schema = {
+            vol.Required(CONF_USERNAME): str,
+            vol.Required(CONF_TOKEN): str,
+        }
```

Comment:
https://github.com/home-assistant/core/blob/50a46933f62c2495cdc5e781ee13aa872c98036b/homeassistant/components/shelly/config_flow.py#L60-L65

Reply:
It is done.